File: mailblastr-go/contacts.go
package mailblastr

import (
	"context"
	"net/http"
	"net/url"
	"strconv"
)

// Contact is a single contact record. Contacts are DOMAIN-FIRST: each sending
// domain has its own contact pool, so the same email address on two domains
// is two records with separate consent.
type Contact struct {
	Object       string `json:"object"`
	Id           string `json:"id"`
	Email        string `json:"email"`
	FirstName    string `json:"first_name"`
	LastName     string `json:"last_name"`
	Unsubscribed bool   `json:"unsubscribed"`
	// Properties are custom contact properties (own values merged over
	// registered fallbacks).
	Properties map[string]any `json:"properties,omitempty"`
	CreatedAt  string         `json:"created_at"`
}

// CreateContactRequest creates a contact. Domain-first: set Domain to create
// in that domain's contact pool via the flat /contacts API (REQUIRED there),
// or AudienceId to target a specific audience via the nested API.
type CreateContactRequest struct {
	// Domain is the sending domain whose contact pool the contact lands in
	// (e.g. "yourdomain.com" — one of your domains). REQUIRED whenever
	// AudienceId is omitted.
	Domain string `json:"domain,omitempty"`
	// AudienceId targets a specific audience via the nested
	// /audiences/:id/contacts API instead of Domain. Not sent in the body.
	AudienceId   string         `json:"-"`
	Email        string         `json:"email"`
	FirstName    string         `json:"first_name,omitempty"`
	LastName     string         `json:"last_name,omitempty"`
	Unsubscribed bool           `json:"unsubscribed,omitempty"`
	Properties   map[string]any `json:"properties,omitempty"`
}

// GetContactRequest identifies a contact for retrieval. Id may be a contact
// id (exact) or an EMAIL; an email can exist in several domains' pools, so
// set Domain to pick the pool (omitted => the oldest match anywhere).
type GetContactRequest struct {
	AudienceId string
	Domain     string
	// Id is a contact id OR email.
	Id string
}

// ListContactsRequest lists contacts. Domain-first: Domain is REQUIRED on the
// flat /contacts list (names the pool) whenever AudienceId is omitted.
type ListContactsRequest struct {
	Domain     string
	AudienceId string
	Limit      int
	After      string
	Before     string
	// SegmentId restricts to members of that segment (works with or without
	// AudienceId).
	SegmentId string
}

// ContactInput is a single contact in a bulk import (no audience — it's on
// the call).
type ContactInput struct {
	Email        string         `json:"email"`
	FirstName    string         `json:"first_name,omitempty"`
	LastName     string         `json:"last_name,omitempty"`
	Unsubscribed bool           `json:"unsubscribed,omitempty"`
	Properties   map[string]any `json:"properties,omitempty"`
}

// BatchContactsRequest bulk-imports contacts from a slice. Upserts by email;
// at most MaxImportContacts per call. Rows missing a valid email are counted
// as skipped rather than failing the request, and Properties are sanitized but
// not validated against the contact-property registry on this endpoint.
type BatchContactsRequest struct {
	AudienceId string
	Contacts   []ContactInput
	// OnConflict "skip" leaves existing contacts untouched (default "upsert").
	OnConflict string
}

// ImportContactsRequest bulk-imports contacts from CSV text (header row
// optional). Upserts by email. By default every non-builtin CSV column is
// auto-registered as a custom property; set CreateProperties to
// mailblastr.Bool(false) for strict mode.
//
// Provide Csv for an inline import (max MaxInlineImportBytes of UTF-8 and
// MaxImportContacts rows), or StorageKey to import a file already uploaded
// via Contacts.ImportUpload.
type ImportContactsRequest struct {
	AudienceId string
	Csv        string
	// FileName labels the archived source file (default "contacts.csv").
	FileName string
	// StorageKey imports a file uploaded through Contacts.ImportUpload
	// instead of inline Csv text.
	StorageKey string
	// OnConflict "skip" leaves existing contacts untouched (default "upsert").
	OnConflict       string
	CreateProperties *bool
	// SegmentId adds every imported address to that segment. It must be one
	// of your segments and belong to this audience.
	SegmentId string
}

// ImportContactsUploadRequest mints a direct-upload URL for a CSV too large to
// send inline.
type ImportContactsUploadRequest struct {
	AudienceId string `json:"-"`
	// Filename must end in ".csv".
	Filename string `json:"filename"`
	// Size is the file size in bytes; must be > 0 and at most
	// MaxContactImportUploadBytes.
	Size int64 `json:"size"`
}

// ImportContactsUploadResponse carries a presigned direct-upload URL. UploadUrl
// is a bearer credential — do not log it. PUT the CSV bytes there with the
// given ContentType, then pass StorageKey to Contacts.Import.
type ImportContactsUploadResponse struct {
	Object      string `json:"object"`
	StorageKey  string `json:"storage_key"`
	UploadUrl   string `json:"upload_url"`
	ContentType string `json:"content_type"`
	// ExpiresIn is the URL's lifetime in seconds.
	ExpiresIn int   `json:"expires_in"`
	MaxBytes  int64 `json:"max_bytes"`
}

// ImportSourceFile describes the archived copy of an imported CSV.
type ImportSourceFile struct {
	FileName   string `json:"file_name"`
	StorageKey string `json:"storage_key"`
	Archived   bool   `json:"archived"`
}

// ImportContactLimit reports how the account's contact allowance shaped a CSV
// import.
type ImportContactLimit struct {
	Plan            map[string]any `json:"plan"`
	UsedBefore      int            `json:"used_before"`
	Limit           int            `json:"limit"`
	RemainingBefore *int           `json:"remaining_before"`
	RemainingAfter  *int           `json:"remaining_after"`
	LimitSkipped    int            `json:"limit_skipped"`
	Reached         bool           `json:"reached"`
	Message         string         `json:"message"`
}

// Contact-import limits enforced by the API.
const (
	// MaxImportContacts caps rows per Contacts.Batch, Contacts.Import (inline)
	// and Audiences.ImportSheet call.
	MaxImportContacts = 10000
	// MaxInlineImportBytes caps the UTF-8 size of inline CSV text.
	MaxInlineImportBytes = 5 * 1024 * 1024
	// MaxContactImportUploadBytes caps a direct CSV upload.
	MaxContactImportUploadBytes = 256 * 1024 * 1024
)

// ImportContactsResponse is the result of a batch / CSV contact import. The
// fields below Total are populated by the CSV import only.
type ImportContactsResponse struct {
	Object string `json:"object"`
	// Imported counts newly inserted contacts.
	Imported int `json:"imported"`
	// Updated counts existing contacts updated.
	Updated int `json:"updated"`
	// Skipped counts rows dropped: missing/invalid email OR left untouched
	// under OnConflict "skip".
	Skipped int `json:"skipped"`
	// Total counts distinct contacts processed.
	Total int `json:"total"`

	// InvalidRows counts CSV rows without a usable email.
	InvalidRows int `json:"invalid_rows,omitempty"`
	// LimitSkipped counts rows dropped because the plan's contact allowance
	// ran out mid-import.
	LimitSkipped int `json:"limit_skipped,omitempty"`
	// SystemSkipped counts rows the importer rejected internally.
	SystemSkipped int `json:"system_skipped,omitempty"`
	// IgnoredColumns lists CSV headers that were not imported.
	IgnoredColumns []string `json:"ignored_columns,omitempty"`
	// SegmentAdded is present only when SegmentId was supplied.
	SegmentAdded int                 `json:"segment_added,omitempty"`
	SourceFile   *ImportSourceFile   `json:"source_file,omitempty"`
	ContactLimit *ImportContactLimit `json:"contact_limit,omitempty"`
}

// UpdateContactRequest updates a contact.
type UpdateContactRequest struct {
	// Domain disambiguates an EMAIL Id across domains (a contact id is exact
	// and needs no domain). Flat API only.
	Domain string `json:"domain,omitempty"`
	// AudienceId is the audience the contact belongs to. OMIT for the flat
	// /contacts/:id API. Not sent in the body.
	AudienceId string `json:"-"`
	// Id is a contact id OR email. Not sent in the body.
	Id           string         `json:"-"`
	FirstName    *string        `json:"first_name,omitempty"`
	LastName     *string        `json:"last_name,omitempty"`
	Unsubscribed *bool          `json:"unsubscribed,omitempty"`
	Properties   map[string]any `json:"properties,omitempty"`
}

// RemoveContactRequest identifies a contact for deletion. On the flat API,
// set Domain when Id is an EMAIL (disambiguates across pools).
type RemoveContactRequest struct {
	AudienceId string
	Domain     string
	// Id is a contact id OR email.
	Id string
}

// RemoveContactResponse is returned by contact deletion.
type RemoveContactResponse struct {
	Object string `json:"object"`
	// Id is the deleted contact's id.
	Id      string `json:"id"`
	Deleted bool   `json:"deleted"`

	// Contact is always empty: the API returns the deleted contact's id as
	// `id`, never as `contact`.
	//
	// Deprecated: read Id instead.
	Contact string `json:"contact,omitempty"`
}

// SegmentMembershipRemoved is returned when removing a contact from a segment.
type SegmentMembershipRemoved struct {
	// Id is the contact id.
	Id string `json:"id"`
	// AudienceId carries the SEGMENT id despite its name — that is what the
	// API returns, camelCase key and all.
	AudienceId string `json:"audienceId"`
	Deleted    bool   `json:"deleted"`
}

// ContactSegmentRef is a segment reference as returned by
// ContactsService.ListSegments (GET /contacts/:id/segments). The route returns
// only Id, Name and CreatedAt — not the full Segment. Use Segments.Get for the
// audience, filter and timestamps.
type ContactSegmentRef struct {
	Id        string `json:"id"`
	Name      string `json:"name"`
	CreatedAt string `json:"created_at"`
}

// ContactTopicSubscription is one topic subscription state of a contact.
type ContactTopicSubscription struct {
	Id          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
	// Subscription is "opt_in" | "opt_out".
	Subscription string `json:"subscription"`
}

// ContactTopics is a contact's topic subscription list.
type ContactTopics struct {
	Object  string                     `json:"object"`
	HasMore bool                       `json:"has_more"`
	Data    []ContactTopicSubscription `json:"data"`
}

// TopicSubscriptionUpdate sets one topic's subscription state.
type TopicSubscriptionUpdate struct {
	Id string `json:"id"`
	// Subscription is "opt_in" | "opt_out".
	Subscription string `json:"subscription"`
}

// UpdateContactTopicsRequest updates a contact's topic subscriptions.
type UpdateContactTopicsRequest struct {
	Topics []TopicSubscriptionUpdate `json:"topics"`
}

// ContactsService handles the flat /contacts and nested
// /audiences/:id/contacts endpoints.
type ContactsService struct {
	client *Client
}

// Create creates a contact; returns the slim ack { object, id }.
// POST /contacts (flat, Domain required) or POST /audiences/:id/contacts.
func (s *ContactsService) Create(params *CreateContactRequest) (*ObjectRef, error) {
	return s.CreateWithContext(context.Background(), params)
}

// CreateWithContext creates a contact.
func (s *ContactsService) CreateWithContext(ctx context.Context, params *CreateContactRequest) (*ObjectRef, error) {
	body := *params
	if params.AudienceId != "" {
		// The nested audience route derives its pool from the path; only the
		// flat route takes domain in the body.
		body.Domain = ""
		return request[ObjectRef](ctx, s.client, http.MethodPost, "/audiences/"+esc(params.AudienceId)+"/contacts", &body, nil)
	}
	return request[ObjectRef](ctx, s.client, http.MethodPost, "/contacts", &body, nil)
}

// Get retrieves a contact by id or email. An id is exact; an EMAIL can exist
// in several domains' pools, so set Domain to pick the pool.
func (s *ContactsService) Get(params *GetContactRequest) (*Contact, error) {
	return s.GetWithContext(context.Background(), params)
}

// GetWithContext retrieves a contact by id or email.
func (s *ContactsService) GetWithContext(ctx context.Context, params *GetContactRequest) (*Contact, error) {
	if params.AudienceId != "" {
		return request[Contact](ctx, s.client, http.MethodGet, "/audiences/"+esc(params.AudienceId)+"/contacts/"+esc(params.Id), nil, nil)
	}
	path := "/contacts/" + esc(params.Id)
	if params.Domain != "" {
		path += "?domain=" + url.QueryEscape(params.Domain)
	}
	return request[Contact](ctx, s.client, http.MethodGet, path, nil, nil)
}

// List lists contacts. Domain-first: Domain is required on the flat /contacts
// list (names the pool) whenever AudienceId is omitted.
func (s *ContactsService) List(params *ListContactsRequest) (*ListResponse[Contact], error) {
	return s.ListWithContext(context.Background(), params)
}

// ListWithContext lists contacts.
func (s *ContactsService) ListWithContext(ctx context.Context, params *ListContactsRequest) (*ListResponse[Contact], error) {
	base := "/contacts"
	q := url.Values{}
	if params != nil {
		if params.AudienceId != "" {
			base = "/audiences/" + esc(params.AudienceId) + "/contacts"
		} else if params.Domain != "" {
			q.Set("domain", params.Domain)
		}
		if params.Limit > 0 {
			q.Set("limit", strconv.Itoa(params.Limit))
		}
		if params.After != "" {
			q.Set("after", params.After)
		}
		if params.Before != "" {
			q.Set("before", params.Before)
		}
		// segment_id filter is honored by both the flat and audience-scoped list.
		if params.SegmentId != "" {
			q.Set("segment_id", params.SegmentId)
		}
	}
	path := base
	if enc := q.Encode(); enc != "" {
		path += "?" + enc
	}
	return request[ListResponse[Contact]](ctx, s.client, http.MethodGet, path, nil, nil)
}

// Batch bulk-imports contacts from a slice. Upserts by email; max 10,000 per
// call. POST /audiences/:id/contacts/batch
func (s *ContactsService) Batch(params *BatchContactsRequest) (*ImportContactsResponse, error) {
	return s.BatchWithContext(context.Background(), params)
}

// BatchWithContext bulk-imports contacts from a slice.
func (s *ContactsService) BatchWithContext(ctx context.Context, params *BatchContactsRequest) (*ImportContactsResponse, error) {
	path := "/audiences/" + esc(params.AudienceId) + "/contacts/batch"
	if params.OnConflict != "" {
		path += "?on_conflict=" + url.QueryEscape(params.OnConflict)
	}
	body := map[string]any{"contacts": params.Contacts}
	return request[ImportContactsResponse](ctx, s.client, http.MethodPost, path, body, nil)
}

// Import bulk-imports contacts from CSV text. Upserts by email.
// POST /audiences/:id/contacts/import
func (s *ContactsService) Import(params *ImportContactsRequest) (*ImportContactsResponse, error) {
	return s.ImportWithContext(context.Background(), params)
}

// ImportWithContext bulk-imports contacts from CSV text.
func (s *ContactsService) ImportWithContext(ctx context.Context, params *ImportContactsRequest) (*ImportContactsResponse, error) {
	q := url.Values{}
	if params.OnConflict != "" {
		q.Set("on_conflict", params.OnConflict)
	}
	if params.CreateProperties != nil && !*params.CreateProperties {
		q.Set("create_properties", "false")
	}
	if params.SegmentId != "" {
		q.Set("segment_id", params.SegmentId)
	}
	path := "/audiences/" + esc(params.AudienceId) + "/contacts/import"
	if enc := q.Encode(); enc != "" {
		path += "?" + enc
	}
	body := map[string]any{}
	if params.StorageKey != "" {
		body["storage_key"] = params.StorageKey
	} else {
		body["csv"] = params.Csv
	}
	if params.FileName != "" {
		body["file_name"] = params.FileName
	}
	return request[ImportContactsResponse](ctx, s.client, http.MethodPost, path, body, nil)
}

// ImportUpload mints a presigned direct-upload URL for a CSV too large to send
// inline. Upload the bytes to the returned UploadUrl, then call Import with
// the returned StorageKey. POST /audiences/:id/contacts/import/upload
func (s *ContactsService) ImportUpload(params *ImportContactsUploadRequest) (*ImportContactsUploadResponse, error) {
	return s.ImportUploadWithContext(context.Background(), params)
}

// ImportUploadWithContext mints a presigned direct-upload URL for a CSV.
// POST /audiences/:id/contacts/import/upload
func (s *ContactsService) ImportUploadWithContext(ctx context.Context, params *ImportContactsUploadRequest) (*ImportContactsUploadResponse, error) {
	path := "/audiences/" + esc(params.AudienceId) + "/contacts/import/upload"
	return request[ImportContactsUploadResponse](ctx, s.client, http.MethodPost, path, params, nil)
}

// Update updates a contact; returns the slim ack { object, id }. On the flat
// API, set Domain when Id is an EMAIL (disambiguates across pools).
func (s *ContactsService) Update(params *UpdateContactRequest) (*ObjectRef, error) {
	return s.UpdateWithContext(context.Background(), params)
}

// UpdateWithContext updates a contact.
func (s *ContactsService) UpdateWithContext(ctx context.Context, params *UpdateContactRequest) (*ObjectRef, error) {
	body := *params
	if params.AudienceId != "" {
		// The nested audience route derives its pool from the path.
		body.Domain = ""
		return request[ObjectRef](ctx, s.client, http.MethodPatch, "/audiences/"+esc(params.AudienceId)+"/contacts/"+esc(params.Id), &body, nil)
	}
	return request[ObjectRef](ctx, s.client, http.MethodPatch, "/contacts/"+esc(params.Id), &body, nil)
}

// Remove deletes a contact. On the flat API, set Domain when Id is an EMAIL.
func (s *ContactsService) Remove(params *RemoveContactRequest) (*RemoveContactResponse, error) {
	return s.RemoveWithContext(context.Background(), params)
}

// RemoveWithContext deletes a contact.
func (s *ContactsService) RemoveWithContext(ctx context.Context, params *RemoveContactRequest) (*RemoveContactResponse, error) {
	if params.AudienceId != "" {
		return request[RemoveContactResponse](ctx, s.client, http.MethodDelete, "/audiences/"+esc(params.AudienceId)+"/contacts/"+esc(params.Id), nil, nil)
	}
	path := "/contacts/" + esc(params.Id)
	if params.Domain != "" {
		path += "?domain=" + url.QueryEscape(params.Domain)
	}
	return request[RemoveContactResponse](ctx, s.client, http.MethodDelete, path, nil, nil)
}

// AddToSegment adds a contact to a segment; returns { id } (the segment id).
// POST /contacts/:id/segments/:segmentId
func (s *ContactsService) AddToSegment(id, segmentId string) (*IdResponse, error) {
	return s.AddToSegmentWithContext(context.Background(), id, segmentId)
}

// AddToSegmentWithContext adds a contact to a segment.
func (s *ContactsService) AddToSegmentWithContext(ctx context.Context, id, segmentId string) (*IdResponse, error) {
	return request[IdResponse](ctx, s.client, http.MethodPost, "/contacts/"+esc(id)+"/segments/"+esc(segmentId), nil, nil)
}

// RemoveFromSegment removes a contact from a segment.
// DELETE /contacts/:id/segments/:segmentId
func (s *ContactsService) RemoveFromSegment(id, segmentId string) (*SegmentMembershipRemoved, error) {
	return s.RemoveFromSegmentWithContext(context.Background(), id, segmentId)
}

// RemoveFromSegmentWithContext removes a contact from a segment.
func (s *ContactsService) RemoveFromSegmentWithContext(ctx context.Context, id, segmentId string) (*SegmentMembershipRemoved, error) {
	return request[SegmentMembershipRemoved](ctx, s.client, http.MethodDelete, "/contacts/"+esc(id)+"/segments/"+esc(segmentId), nil, nil)
}

// ListSegments lists the segments a contact belongs to. Rows are the reduced
// ContactSegmentRef shape (Id, Name, CreatedAt), not the full Segment — use
// Segments.Get for that. Passing nil params returns every segment: this
// endpoint only pages when you supply pagination params.
// GET /contacts/:id/segments
func (s *ContactsService) ListSegments(id string, params *ListParams) (*ListResponse[ContactSegmentRef], error) {
	return s.ListSegmentsWithContext(context.Background(), id, params)
}

// ListSegmentsWithContext lists the segments a contact belongs to.
func (s *ContactsService) ListSegmentsWithContext(ctx context.Context, id string, params *ListParams) (*ListResponse[ContactSegmentRef], error) {
	return request[ListResponse[ContactSegmentRef]](ctx, s.client, http.MethodGet, listPath("/contacts/"+esc(id)+"/segments", params), nil, nil)
}

// GetTopics gets a contact's topic subscriptions. GET /contacts/:id/topics
func (s *ContactsService) GetTopics(id string) (*ContactTopics, error) {
	return s.GetTopicsWithContext(context.Background(), id)
}

// GetTopicsWithContext gets a contact's topic subscriptions.
func (s *ContactsService) GetTopicsWithContext(ctx context.Context, id string) (*ContactTopics, error) {
	return request[ContactTopics](ctx, s.client, http.MethodGet, "/contacts/"+esc(id)+"/topics", nil, nil)
}

// UpdateTopics updates a contact's topic subscriptions; returns { id } (the
// contact id). PATCH /contacts/:id/topics
func (s *ContactsService) UpdateTopics(id string, params *UpdateContactTopicsRequest) (*IdResponse, error) {
	return s.UpdateTopicsWithContext(context.Background(), id, params)
}

// UpdateTopicsWithContext updates a contact's topic subscriptions.
func (s *ContactsService) UpdateTopicsWithContext(ctx context.Context, id string, params *UpdateContactTopicsRequest) (*IdResponse, error) {
	return request[IdResponse](ctx, s.client, http.MethodPatch, "/contacts/"+esc(id)+"/topics", params, nil)
}

File: mailblastr-python/mailblastr/contacts.py
"""Contacts resource — DOMAIN-FIRST flat /contacts API plus the nested
/audiences/:id/contacts variants, bulk imports, segment membership and topics.

Domain-first model: each sending domain has its own contact POOL — the same
address on two domains is two records with separate consent. On the flat
``/contacts`` API, ``domain`` is REQUIRED to create/list and disambiguates an
EMAIL id on get/update/remove. Pass ``audience_id`` instead to target a
specific audience via the nested API.
"""

from typing import Dict, List, TypedDict, Union

from . import http_client
from .http_client import build_query, paginate, path_escape as _e


class CreateParams(TypedDict, total=False):
    domain: str  # REQUIRED on the flat /contacts API (when audience_id is omitted)
    audience_id: str  # target a specific audience via the nested API instead
    email: str  # required
    first_name: str
    last_name: str
    unsubscribed: bool
    properties: Dict[str, Union[str, int, float]]


class UpdateParams(TypedDict, total=False):
    id: str  # required: contact id OR email
    domain: str  # disambiguates an EMAIL id across domains (flat API)
    audience_id: str  # omit for the flat /contacts/:id API
    first_name: str
    last_name: str
    unsubscribed: bool
    properties: Dict[str, Union[str, int, float]]


class ContactInput(TypedDict, total=False):
    email: str  # required
    first_name: str
    last_name: str
    unsubscribed: bool
    properties: Dict[str, Union[str, int, float]]


class UpdateTopicsParams(TypedDict):
    topics: List[Dict[str, str]]  # [{"id": ..., "subscription": "opt_in" | "opt_out"}]


class ImportCsvParams(TypedDict, total=False):
    audience_id: str  # required
    csv: str  # inline CSV text (max 5 MB, 10,000 rows) — or pass storage_key
    storage_key: str  # key from import_upload(), for files too big to inline
    file_name: str  # archived name for inline CSV; defaults to contacts.csv
    on_conflict: str  # 'upsert' (default) | 'skip'
    create_properties: bool  # False -> don't auto-register unseen columns
    segment_id: str  # add every imported email to this segment


class ImportUploadParams(TypedDict, total=False):
    audience_id: str  # required
    filename: str  # required, must end in .csv
    size: int  # required, bytes; max 256 MB


class Contacts:
    """``mailblastr.Contacts`` — the /contacts (+ nested audience) endpoints."""

    CreateParams = CreateParams
    UpdateParams = UpdateParams
    ContactInput = ContactInput
    UpdateTopicsParams = UpdateTopicsParams
    ImportCsvParams = ImportCsvParams
    ImportUploadParams = ImportUploadParams

    @classmethod
    def create(cls, params):
        """Create a contact. Domain-first: pass ``domain`` to create in that
        domain's contact pool via the flat ``/contacts`` API (required there),
        or ``audience_id`` to target a specific audience via the nested API."""
        body = dict(params)
        audience_id = body.pop("audience_id", None)
        domain = body.pop("domain", None)
        if audience_id:
            # The nested audience route derives its pool from the path.
            return http_client.request("POST", f"/audiences/{_e(audience_id)}/contacts", body)
        if domain is not None:
            body["domain"] = domain
        return http_client.request("POST", "/contacts", body)

    @classmethod
    def get(cls, params):
        """Retrieve a contact by id or email — ``{"id", "domain"?, "audience_id"?}``.
        An id is exact; an EMAIL can exist in several domains' pools, so pass
        ``domain`` to pick the pool (omitted → the oldest match anywhere)."""
        cid = _e(params["id"])
        audience_id = params.get("audience_id")
        if audience_id:
            return http_client.request("GET", f"/audiences/{_e(audience_id)}/contacts/{cid}")
        qs = build_query({"domain": params.get("domain")})
        return http_client.request("GET", f"/contacts/{cid}{qs}")

    @classmethod
    def list(cls, params=None):
        """List contacts. Domain-first: ``domain`` is required on the flat
        ``/contacts`` list (names the pool); or pass ``audience_id`` for the
        nested list. ``segment_id`` filters either variant."""
        params = params or {}
        audience_id = params.get("audience_id")
        qs = build_query(
            {
                "domain": params.get("domain") if not audience_id else None,
                "limit": params.get("limit"),
                "after": params.get("after"),
                "before": params.get("before"),
                "segment_id": params.get("segment_id"),
            }
        )
        base = f"/audiences/{_e(audience_id)}/contacts" if audience_id else "/contacts"
        return http_client.request("GET", f"{base}{qs}")

    @classmethod
    def batch(cls, params):
        """Bulk-import contacts from a list — ``{"audience_id", "contacts",
        "on_conflict"?}``. Upserts by email; max 10,000 per call.
        ``on_conflict='skip'`` leaves existing contacts untouched.

        Rows without a valid ``email`` are counted as ``skipped`` rather than
        failing the request, and ``properties`` are NOT checked against the
        contact-property registry here (unlike the single-contact writes).
        POST /audiences/:id/contacts/batch"""
        qs = build_query({"on_conflict": params.get("on_conflict")})
        return http_client.request(
            "POST",
            f"/audiences/{_e(params['audience_id'])}/contacts/batch{qs}",
            {"contacts": params["contacts"]},
        )

    @classmethod
    def import_csv(cls, params):
        """Bulk-import contacts from CSV. Upserts by email. By default every
        non-builtin CSV column is auto-registered as a custom property; pass
        ``create_properties=False`` for strict mode. ``segment_id`` adds every
        imported email to that segment. POST /audiences/:id/contacts/import

        Pass ``csv`` for inline text (max 5 MB and 10,000 rows), or a
        ``storage_key`` from :meth:`import_upload` for anything larger."""
        qs = build_query(
            {
                "on_conflict": params.get("on_conflict"),
                "create_properties": (
                    "false" if params.get("create_properties") is False else None
                ),
                "segment_id": params.get("segment_id"),
            }
        )
        body = {}
        for key in ("csv", "storage_key", "file_name"):
            if params.get(key) is not None:
                body[key] = params[key]
        return http_client.request(
            "POST", f"/audiences/{_e(params['audience_id'])}/contacts/import{qs}", body
        )

    @classmethod
    def import_upload(cls, params):
        """Mint a presigned direct-upload URL for a CSV too large to inline
        (up to 256 MB) — ``{"audience_id", "filename", "size"}``. PUT the file
        to the returned ``upload_url``, then pass the returned ``storage_key``
        to :meth:`import_csv`. POST /audiences/:id/contacts/import/upload

        The ``upload_url`` is a bearer credential — do not log it."""
        return http_client.request(
            "POST",
            f"/audiences/{_e(params['audience_id'])}/contacts/import/upload",
            {"filename": params["filename"], "size": params["size"]},
        )

    @classmethod
    def update(cls, params):
        """Update a contact. On the flat API, pass ``domain`` when ``id`` is
        an EMAIL (disambiguates across pools). PATCH /contacts/:id"""
        body = dict(params)
        cid = _e(body.pop("id"))
        audience_id = body.pop("audience_id", None)
        domain = body.pop("domain", None)
        if audience_id:
            return http_client.request(
                "PATCH", f"/audiences/{_e(audience_id)}/contacts/{cid}", body
            )
        if domain is not None:
            body["domain"] = domain
        return http_client.request("PATCH", f"/contacts/{cid}", body)

    @classmethod
    def remove(cls, params):
        """Delete a contact — ``{"id", "domain"?, "audience_id"?}``. On the
        flat API, pass ``domain`` when ``id`` is an EMAIL."""
        cid = _e(params["id"])
        audience_id = params.get("audience_id")
        if audience_id:
            return http_client.request("DELETE", f"/audiences/{_e(audience_id)}/contacts/{cid}")
        qs = build_query({"domain": params.get("domain")})
        return http_client.request("DELETE", f"/contacts/{cid}{qs}")

    @classmethod
    def add_to_segment(cls, contact_id, segment_id):
        """Add a contact to a segment. POST /contacts/:id/segments/:segment_id"""
        return http_client.request(
            "POST", f"/contacts/{_e(contact_id)}/segments/{_e(segment_id)}"
        )

    @classmethod
    def remove_from_segment(cls, contact_id, segment_id):
        """Remove a contact from a segment. DELETE /contacts/:id/segments/:segment_id"""
        return http_client.request(
            "DELETE", f"/contacts/{_e(contact_id)}/segments/{_e(segment_id)}"
        )

    @classmethod
    def list_segments(cls, contact_id, params=None):
        """List the segments a contact belongs to (``id``/``name``/
        ``created_at`` only, not full segment objects). Called with no
        pagination params every segment is returned.
        GET /contacts/:id/segments"""
        return http_client.request(
            "GET", f"/contacts/{_e(contact_id)}/segments{paginate(params)}"
        )

    @classmethod
    def get_topics(cls, contact_id, params=None):
        """Get a contact's topic subscriptions. Called with no pagination
        params every topic is returned. GET /contacts/:id/topics"""
        return http_client.request(
            "GET", f"/contacts/{_e(contact_id)}/topics{paginate(params)}"
        )

    @classmethod
    def update_topics(cls, contact_id, params):
        """Update a contact's topic subscriptions. PATCH /contacts/:id/topics"""
        return http_client.request("PATCH", f"/contacts/{_e(contact_id)}/topics", params)
